pkg/addon/push.go

```diff
@@ -0,0 +1,296 @@
+type PushCmd struct {
+	ChartName          string
+	AppVersion         string
+	ChartVersion       string
+	RepoName           string
+	Username           string
+	Password           string
+	AccessToken        string
+	AuthHeader         string
+	ContextPath        string
+	ForceUpload        bool
+	UseHTTP            bool
+	CaFile             string
+	CertFile           string
+	KeyFile            string
+	InsecureSkipVerify bool
+	Out                io.Writer
+	Timeout            int64
+	// We need it to search in addon registries.
+	// If you use URL, instead of registry names, then it is not needed.
+	Client client.Client
+}
+
+// Push pushes addons (i.e. Helm Charts) to ChartMuseum.
+// It will package the addon into a Helm Chart if necessary.
+func (p *PushCmd) Push(ctx context.Context) error {
+	var repo *cmhelm.Repo
+	var err error
+
+	// Get the user specified Helm repo
+	repo, err = GetHelmRepo(ctx, p.Client, p.RepoName)
+	if err != nil {
+		return err
+	}
+
+	// Make the addon dir a Helm Chart (if not already)
+	err = MakeChart(p.ChartName)
+	// `Not a directory` errors are ignored, that's fine,
+	// since .tgz files are also supported.
+	if err != nil && !strings.Contains(err.Error(), "is not a directory") {
+		return err
+	}
+
+	// Get chart from a directory or .tgz package
+	chart, err := cmhelm.GetChartByName(p.ChartName)
+	if err != nil {
+		return err
+	}
+
+	// Override chart version using specified version
+	if p.ChartVersion != "" {
+		chart.SetVersion(p.ChartVersion)
+	}
+
+	// Override app version using specified version
+	if p.AppVersion != "" {
+		chart.SetAppVersion(p.AppVersion)
+	}
+
+	// Override username and password using specified values
+	username := repo.Config.Username
+	password := repo.Config.Password
+	if p.Username != "" {
+		username = p.Username
+	}
+	if p.Password != "" {
+		password = p.Password
+	}
+
+	// Unset accessToken if repo credentials are provided
+	if username != "" && password != "" {
+		p.AccessToken = ""
+	}
+
+	// In case the repo is stored with cm:// protocol,
+	// (if that's somehow possible with KubeVela addon registries)
+	// use http instead,
+	// otherwise keep as it-is.
+	var url string
+	if p.UseHTTP {
+		url = strings.Replace(repo.Config.URL, "cm://", "http://", 1)
+	} else {
+		url = strings.Replace(repo.Config.URL, "cm://", "https://", 1)
+	}
+
+	cmClient, err := cm.NewClient(
+		cm.URL(url),
+		cm.Username(username),
+		cm.Password(password),
+		cm.AccessToken(p.AccessToken),
+		cm.AuthHeader(p.AuthHeader),
+		cm.ContextPath(p.ContextPath),
+		cm.CAFile(p.CaFile),
+		cm.CertFile(p.CertFile),
+		cm.KeyFile(p.KeyFile),
+		cm.InsecureSkipVerify(p.InsecureSkipVerify),
+		cm.Timeout(p.Timeout),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	// Use a temporary dir to hold packaged .tgz Charts
+	tmp, err := ioutil.TempDir("", "helm-push-")
+	if err != nil {
+		return err
+	}
+	defer func(path string) {
+		_ = os.RemoveAll(path)
```

Comment:
when user specify a tgz file to push, will this tgz be deleted?

Reply:
If the use specifies a .tgz file, it is loaded into memory first (because we need to override some fields, if the user wants), then it is written to a temporary path. So, it deletes the file in the temporary path.

The original file is not touched.

Reply:
good